### ISSUE TITLE
Alamode development setup

### ISSUE THREAD

**Issue by kerber** [RaspberryPints/RaspberryPints#194]
Hey everyone,

I've gone ahead and committed the Arduino monitoring code to the 'feature-flow_monitoring' branch.

To get RPi set up for the Alamode, follow [Wyolum's documentation](https://docs.google.com/document/d/1HBvd3KNmcs632ZgO6t_u37B-qwV6P9o9FQe62lGkumM).

Since the flow sensors will be pulling power, we want to bypass the PRi's onboard fuse. To do this, set your 5V_Link jumper to ON. Then, plug the power into the Alamode board instead of the RPi.

**Comment by kerber** [comment]
Note that this current setup excludes the Centipede board. While it gives us a lot of expandability for future development, it's an unnecessary level of complexity for simple keg monitoring.

Currently, for testing sake, wire the flow sensors to the Alamode as follows:
- Pin1(red) -> Ground
- Pin2 -> Any digital pin 2-13. Be sure to set this at the top of the Arduino code
- Pin3 -> VIN

Now run the following to get python to support serial

``` bash
sudo apt-get install python-serial
```

Until we get the API and necessary changes made, the python does not integrate with the web front-end. However, if you run flow_monitor.py from the terminal, it will print out any messages sent from the Arduino.

**Comment by jkurl** [comment]
Man my sink in my stateroom on the ship is gonna get a workout testing the
flow meter feature. Just remember everyone to pay more taxes incase I break
my sink and need a new one. The govt is broke. :p

**Comment by thadius856** [comment]
You JERK! I was about to get a few minutes of video games in while I'm waiting for the boil to start, and now you post this. >:\

Ugh. Warming up my soldering iron now...

**Comment by mcangeli** [comment]
Ok... so this is to the alamode directly and not the centipede?

**Comment by kerber** [comment]
Correct. I've tested with the MCP23017 quite a bit, so I wanted to scale back and remove a (currently) unnecessary level of complexity.

**Comment by kerber** [comment]
Without the overhead of the i2c protocol, we actually get better precision without the Centipede.

**Comment by mcangeli** [comment]
Question related to the flowmeters and not the alamode...

How are you all planning to connect the tap tubing to the flowmeter??

Has anyone had any luck just sliding the tubing in to the meter or are you using something like: http://store.kegbot.org/products/barb-connector-pair-1-4-barb ??

**Comment by jkurl** [comment]
You need the barbs.

**Comment by kerber** [comment]
There was a bit of a discussion in the original thread: http://www.homebrewtalk.com/f51/feeler-any-interest-digital-tap-lists-450400/index21.html#post5802880

**Comment by mcangeli** [comment]
Thanks. I knew I had seen it somewhere before... :)

**Comment by thadius856** [comment]
Hate barbs.

I'd really recommend the stems if you have the choice.

**Comment by mcangeli** [comment]
Ok, on to my next question... (which may be my last since no one has the damned barbs needed...)

Outside of changing the number of meters and the pin id in the arduino code, what else has to be setup?

Does the page automatically hit the python script?

**Comment by thadius856** [comment]
The barbs are for sale on FreshWaterSystems.com

I still recommend reducers/stems over barbs.

**Comment by mcangeli** [comment]
The ones on Freshwatersystems said they were out of stock until 3/14 yesterday. Today it said 3/15 now it says ready to ship. The site needs to make up its mind.

I'm not putting a lot of faith in that.

Grainger sells stems by the 10pk, but not in the 3/8 to 5/16 size.

I found another site, but the only shipping options are ups at $7 a pop (for a $4 order) or the kegbot site where they're double what I'd pay elsewhere with $8 shipping.

**Comment by kerber** [comment]
As of now, there's no integration between the python code and the webpage. We'll need to get the API and the rest of the flow meter changes going first. JayUnt and I have been collaborating on this an we'll keep everyone updated as things progress.

That said, you can manually run the flow_monitor.py script from the terminal and it will print out any pour or keg kicked messages sent from the Arduino.

**Comment by mcangeli** [comment]
For those who haven't....

sudo apt-get install python-serial

**Comment by kerber** [comment]
Good call! I'll add it to the top of the issue.

**Comment by mcangeli** [comment]
Ok, so for those of us who have no clue... 

step one: Download the Flow-meter release and place files in the /var/www directory over the current files.
Step two (it helps to have a vnc connection or access to the desktop) with the alamode and the flowmeters hooked up, use the arduino-ide program in the "Electronics" folder, click file open. Browse the file system to /var/www/Arduino and select the RaspberryPints.ino file.

Edit the number of flow meters and the pin number (I now see why Kerber used 8,9, 10) and save.
Click the upload button (this compiles and uploads the code to the arduino itself).

Once thats done, in a cli go to the /var/www/python folder and run: 
python flow-meters.py 

gently blow through the flowmeter in the right direction and you should see output on the screen.

AWESOME WORK KERBER!! Now to get it posting to the db.
![20140224_181748_resized](https://f.cloud.github.com/assets/6456272/2252600/af89d090-9dab-11e3-9c73-44261e883c92.jpg)

**Comment by thadius856** [comment]
Ill try to find time to test this tonight.

**Comment by mcangeli** [comment]
kerber,

couldn't the python script dump the data (the Pin #, the count and whether or not its a kick or a pour) to a table?

I noticed that in the DB theres the pours table and then the vwGetTapsAmountPoured. 

Both of those look to be almost the same thing except that the pours table is to be historical while the vwgettapsamountpoured is strictly for displaying the keg level right?

If we had a poursData table with a max of five fields:

ID (key)
Pin
Count
Status (pour/kick)
Date/time

that the python script dumps that data in to, the php could then handle the math to convert it to the pour and dump that in to the pours table. set a cron to run the php every 5 minutes or so to do that part.

**Comment by thadius856** [comment]
Looks like maybe my Alamode is fried. Hrm.

I think you'd also need a column for the calibration factor on each pour in the table, if I understand his direction properly.

**Comment by mcangeli** [comment]
Thad, have you tried running the blink test on it?

The calibration factor would be in the tap table where the pin is assigned. it wouldn't have to be set for every pour, just every time you tap a new beer (as one might be thicker/lighter than another) if I understand correctly from the conversation he and I had last night.

**Comment by thadius856** [comment]
The thing won't even let me upload to it. Kerb and I went thru everything we could think of last night in IRC.

If the calibration faction changes every keg, then wouldn't old pours get skewed?

**Comment by mcangeli** [comment]
They shouldn’t. I think the way to go would be to have python dump the cycles in to a table then php do the math to convert the cycles to oz (or whatever) and dump that to the pours table.

The math would be something like (C2 – C1)/oz

(C2 is the most recent cycle, c1 is the last one, oz would be ounces per cycle)

It could in theory grab the oz per cycle from the taps table, the c2 and c1 would be the two most recent cycle counts in the pourData table.

Then it dumps that in the pours table.

But I may be way off base from what kerber is thinking.

**Comment by kerber** [comment]
I think we're going to try keep all DB access in PHP. That will keep all DB logic on one place and allow the python to just post to a PHP link. This has a bonus of allowing for instances where people may not want to run the webserver and DB on the RPi, but on a different server. I'm sure there won't be many people looking for that feature, but I definitely will be.

JayUnt and I have been working on a slightly modified DB layout that allows for some added functionality and clearer workflow. We're still hashing out the details, but the info about a specific beer (og, fg, srm, etc) will be stored as a 'batch' of beer. This batch would be linked to the beer, the keg, and the tap. People would then be able to easily swap out partially full kegs without losing all of the information. It also presents a clearer path for batch history. Each pour will contain the time, calibration data of the linked tap at time of pour, the number of pulses, the pin #, and the linked batch. Any other info for the pour can be derived when they are displayed.

Now, the workflow would be something like this... 

You go to your kegs list and have either a 'Kick' button or a 'Fill' button, depending on whether the keg is linked to a batch. If it is linked to a batch and you click 'Kick', then it marks the batch as inactive and changes the keg button to 'Fill'. If you click the 'Fill' button, it gives you a form to fill out the batch specifics and select the 'beer' from a dropdown.

From the taps list, you're able to select from any batch/keg that's not already attached to a tap.

To calibrate, you go to a 'Pour History' page. Each line is a saved pour and displays the time of pour, the volume of pour, the beer, and the tap #. Next to each line is a small 'calibrate' button. On clicking it, you are given 2 fields that are pre-populated with current data. These fields are 'volume' and 'tap #'. 

Now, if you are trying to set up your flow meter pin# -> tap, all you have to do is take a pull from your tap, go into the history, select calibrate, and select the tap. Since the pin# -> tap relation is only stored in the tap table, it just gets updated in there.

If you swap out kegs and want to make sure your calibration is correct, you take a pull into a measuring glass, then go into the history, hit calibrate on the pour, and enter the actual volume of the pour. Since it knows how many pulses the pour was, it calculates the new pulses/liter. It then stores that # with the tap, the pour entry, and any pour entry on that tap since (in case you have taken pours since the measure).

We're still working out some of the details and coming up with a solid schema - as well as an update path. This is going to take some time to work out, but I'm pretty confident that it will be awesome.

**Comment by mcangeli** [comment]
Ok, so you're leaning towards having python out put the pours to a file that the php then parses? (from my basic search that seems to be the recommended way...)

unless its possible to send something like localhost/pourdata.php?reading=P&count=4356&pin=8
thru a get or curl like command.....

**Comment by kerber** [comment]
It would be closer to the latter. 

The reason I'm really interested in doing it this way is to keep all DB logic in the same layer. I'd hate to update the DB, then have to make changes in Python and PHP. We could do it all on the Python side, but it would be a giant PITA to make the calls to Python every time PHP needs data. 

This also allows us to have the python and php on different boxes and do the calls over the network.

**Comment by kerber** [comment]
That said, if anyone has any reasons to do it another method, I'm totally open to it.

**Comment by thadius856** [comment]
I don't have any reasons to even suspect the above is English, so...

**Comment by mcangeli** [comment]
Where in the line are you putting the flowmeter? Closer to the tap or closer to the keg or in the middle of the line? Does it matter?

**Comment by kerber** [comment]
I don't suspect it would matter much. I'm still waiting on my fittings, so I've not been able to hook mine up.

**Comment by thadius856** [comment]
From discussion I saw over at the KegBot project, placement isn't very important.

I'd probably recommend on the QD side, otherwise your shanks will protrude a couple extra inches out and block some kegs from being lifted straight out.

Since I have a dual hinge keezer, I'll mount mine as close to the shanks as possible so they lift out of the way for when replacing kegs.

**Comment by mcangeli** [comment]
I was leaning closer to the shanks and taps as that would be cleaner for
running the wiring.

**Comment by mcangeli** [comment]
Ok... 

a. it helps if you put the flowmeters in line the right direction. hooking them up backwards does no good... ;)

b. it looks like 2005 pulses is a pint glass of water. :beer: 

And the kids like getting the water from the tap as a test.

**Comment by mcangeli** [comment]
Based on one of @kerber 's other projects, I created the init.d script for the flow_monitor.py script.

at the CLI:

```
$ sudo nano /etc/init.d/flowmon
```

and paste the following (make sure to change the directory to wherever you have the python directory for the raspberrypints):

```
#!/bin/sh

### BEGIN INIT INFO
# Provides:          RaspberryPints flow Monitoring
# Required-Start:    $remote_fs $syslog
# Required-Stop:     $remote_fs $syslog
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: Put a short description of the service here
# Description:       Put a long description of the service here
### END INIT INFO

# Change the next 3 lines to suit where you install your script and what you want to call it
DIR=/home/pi/raspberrypints/public/python/
DAEMON=$DIR/flow_monitor.py
DAEMON_NAME=RPFlowMon

# This next line determines what user the script runs as.
# Root generally not recommended but necessary if you are using the Raspberry Pi GPIO from Python.
DAEMON_USER=root

# The process ID of the script when it runs is stored here:
PIDFILE=/var/run/$DAEMON_NAME.pid

. /lib/lsb/init-functions

do_start () {
    log_daemon_msg "Starting system $DAEMON_NAME daemon"
    start-stop-daemon --start --background --pidfile $PIDFILE --make-pidfile --user $DAEMON_USER --chuid $DAEMON_USER --startas $DAEMON
    log_end_msg $?
}
do_stop () {
    log_daemon_msg "Stopping system $DAEMON_NAME daemon"
    start-stop-daemon --stop --pidfile $PIDFILE --retry 10
    log_end_msg $?
}

case "$1" in

    start|stop)
        do_${1}
        ;;

    restart|reload|force-reload)
        do_stop
        do_start
        ;;

    status)
        status_of_proc "$DAEMON_NAME" "$DAEMON" && exit 0 || exit $?
        ;;
    *)
        echo "Usage: /etc/init.d/$DAEMON_NAME {start|stop|restart|status}"
        exit 1
        ;;

esac
exit 0

```

then:

```
$ sudo chmod a+x /etc/init.d/flowmon
$ sudo /etc/init.d/flowmon start
```

**Comment by kerber** [comment]
Excellent!

I'm running one very similar to this and my RaspberryUPS service script: https://github.com/kerber/RaspberryUPS/blob/master/RaspberryUPS

**Comment by mcangeli** [comment]
It should look similar... I grabbed your daemon code from the UPS script and updated it for the flow_monitor.py

;)

**Comment by kerber** [comment]
Right on!

**Comment by mcangeli** [comment]
Also, it looks like there is ~165 pulses per ounce of beer on the sf800.

So a 12oz pour is ~1980 pulses.

**Comment by kerber** [comment]
Since it's rated at 6 pulse per mL, it would be ~180 pulses per oz. 165 sounds totally within reason! I'm looking forward to getting some real world data so we can see just how much variance there can be in the same model. Calibration is going to be such an essential process.

**Comment by jkurl** [comment]
hey guys where is the most up to date code?  I finally got some time to solder my alamode board, hook it up, install all the necessary packages and made sure i could communicate with it.  I want to start testing the flow meter but wasn't sure if kerber was still holding the code or if it has been merged into the main baseline code.  

Sorry for not being around to help test.  We are coming up on the end of a major dockside period and got a 2 week shakedown cruise coming up so my time has been limited on what i get to do at the end of the work day.  let me know where i can help.

Jerry

**Comment by mcangeli** [comment]
The most up to date code.... for the alamode/flow meter stuff is https://github.com/RaspberryPints/RaspberryPints/tree/feature-flow_monitor

The most up to date code for the move to the new format/structure is 
https://github.com/RaspberryPints/RaspberryPints/tree/feature-move-to-laravel

The move to laravel branch does not have any of the flow meter code though.

The flow_monitor code does not currently do anything outside of printing the output of the flow monitor on the screen.

I've been playing with both and have a version of code that is on the new laravel structure (though areas in the admin area are missing/not working) and the flow monitor code (however, its updating the pours table via the python script and not thru php the way kerber envisions it) and it updates both.

I also have a version of the original code with the flow_monitor code working successfully.

You can see what I've been working on in my branches:
https://github.com/mcangeli/RaspberryPints
The untappd tree and the feature_laravel tree

You can see them on my server:
http://rp.skypainterbrewing.com - My "LIVE" site, formatted for my monitor and working with two out of the four flow meters. I'm waiting on four more barbs so I can hook up the other two taps.

http://dev.skypainterbrewing.com - My "DEVELOPMENT" site. This is the site running on the new codebase. It does have the untappd integration (though not to jayunt's satisfaction) and the flow_monitors. This code is under the feature_laravel branch on my repo, but it does not have the alamode and python codes for the flow meters yet ( have to clean up and comment the python).

I'm more than happy to help if you have questions on getting this setup. A big thanks goes to Kerber, Jayunt, Thadius and Boombrewer.....

**Comment by Boombrewer** [comment]
Sorry guys, got busy with work, family, and other projects.  I'm ordering my barbs in the next day or two but it seems like you guys have this pretty much done besides integration with the webpage.

Anyone thought of starting how-to/setup instructions so we'll have something to post for the users to follow.  It sounds pretty straight-forward from what I've read so far.  I know we don't want to post it on the website until we're ready to launch but maybe we can stash it as to "To-Do" item or on the Wiki while we work on it.

Ideas???

**Comment by kerber** [comment]
The majority of the site setup is going to be the same. The flow monitoring process is still early enough dev that it might well change substantially.

**Comment by Boombrewer** [comment]
I was referring to just the flow monitoring process.  I guess I was assuming you guys has hammered out more than you have.

**Comment by kerber** [comment]
Fair enough. We've been primarily focusing on researching and testing the best web technologies to get us where people want RPints to be. While Jay has been kicking his ass migrating the code, I was lounging on vacation. Now that I'm back, I'm digging into it, so dev on the flow monitoring feature should pick back up.

**Comment by Boombrewer** [comment]
Once I get all the parts in and set up, I'll do what I can to help.  It sounds like the python side of the code is pretty solid and complete so I probably won't be much help since PHP still looks like Cyrillic to me.  But I'll do what I can.

**Comment by Skemp45** [comment]
Can anayone post a few pictures of theirs allsetup and wired together?

**Comment by Boombrewer** [comment]
Do you have to modify the alamode as is recommended in Wyolum's documentation (under the Warning section, Hardware Changes)?

Sorry but there seems to be quite a bit of 'progress' from the beginning of this thread to the end and I want to make sure I'm not going to fry my alamode.

Thanks!  Hope you have everything wired up and testing soon.

**Comment by mcangeli** [comment]
@Skemp45 I think this one can be closed?

**Comment by Skemp45** [comment]
Yeah for sure, at a hotel right now for a birthday party can someone else close this please?
